Restructure VID Python/FWLite code and avoid generating dictionaries for `Reco*/*Identification` packages

guitargeek opened the pull request:
#### PR description:

I took issue reported in https://github.com/cms-sw/cmssw/issues/35754 as an opportunity to simplify the VID code a bit on the Python side.

The key class for the Python/FWLite support of VID is the `VersionedSelector<>` in PhysicsTools/SeletorUtils. The dictionaries for the different template instances of this class were split over the Electron/Muon/Photon identification packages. Now, they are all in the `PhysicsTools/SeletorUtils` library.

Then, there were dictionaries in the 3 identification packages for almost trivial classes used in `VIDSelectorBase.py`. However, these classes are tiny template classes that are only used in said python scipt, so I don't think we need dictionaries for them. The classes are small enough such that including them via `gInterpreter.Declare` is very fast.

Finally, there was a dictionary produced for the `ElectronMVAEstimatorRun2` class, which I suggest to also not do anymore. There is only one use case for this: the class is used in the `ElectronIdentification/python/FWLite.py` utility file. But again, declaring the `ElectronMVAEstimatorRun2` is very quick (after avoiding some includes), so it's not necessary to generate dictionaries in CMSSW just for this one usecase (which is outdated anyway because FWLite usage is going down since NanoAOD I guess).

More information can also be found in the commit messages.

#### PR validation:

* CMSSW compiles
* `test_eleid_fwlite.py` runs
* it is still possible to instantiate the Python classes `VIDElectronSelector`, `VIDPhotonSelector`, `VIDMuonSelector`

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport intended.

guitargeek committed:
Restructure VID code for Python/FWLite

* Avoid building dictionaries for tiny helper classes that can also be
  defined with `ROOT.gInterpreter.Declare` on the fly.

* Define dictionaries for `VersionedSelector<>` template instantiations
  in PhysicsTools/SelectorUtils. Like this we don't need to generate
  dictionaries for ElectronIdentification, MuonIdentification, and
  PhotonIdentification, all the dictionaries for VersionedSelectors are
  in one place, and we reduce the dependency of Reco packages on
  PhysicsTools.

guitargeek committed:
Speed up interactive dictionary generation for ElectronMVAEstimatorRun2

Now that the dictionaries for the ElectronIdentification package is
removed, there is still one usecase for which we need to find an
alternative solution. The `ElectronMVAEstimatorRun2` class needs to be
usable from Python in `ElectronIdentification/python/FWLite.py`,
which we can achieved by including `ElectronMVAEstimatorRun2.h` with the
ROOT interpreter in Python. However, doing so was very slow because of
all the includes in `ElectronMVAEstimatorRun2.h` and
`AnyMVAEstimatorRun2Base.h`, and there were even errors caused by
conflicting modules on my system.

Hence, most include statements in `ElectronMVAEstimatorRun2.h` were
replaced by forward declarations in this commit. Now the call to
`gInterpreter` returns almost instantaneously.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/26361

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @guitargeek (Jonas Rembser) for master.

It involves the following packages:

- RecoEgamma/EgammaTools (**reconstruction**)
- RecoEgamma/ElectronIdentification (**reconstruction**)
- RecoEgamma/PhotonIdentification (**reconstruction**)


@jpata, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@Sam-Harper, @jainshilpi, @lgray, @sobhatta, @afiqaize, @wrtabb, @varuns23, @ram1123 this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/26362

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #35953 was updated. @jpata, @cmsbuild, @slava77 can you please check and sign again.

guitargeek committed:
Don't define `thread_safety_macros` in ROOT interpreter

From the ROOT guide: https://root.cern.ch/root/htmldoc/guides/users-guide/Cling.html

"The preprocessor symbol __CLING__ is defined for both ROOT and rootcling.
The symbol __ROOTCLING__ (and __MAKECINT__ for backward compatibility)
is only defined in rootcling."

So to disable the `thread_safety_macros` in an interactive ROOT session too,
we need to check for `__CLING__` and not `__ROOTCLING__`.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/26367

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #35953 was updated. @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @slava77, @jpata can you please check and sign again.

guitargeek commented:
Sorry, I should have made the PR a draft before. I'm done with the force pushes now

makortel commented on PhysicsTools/SelectorUtils/src/classes.h:
```diff
@@ -10,14 +10,28 @@
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/PatCandidates/interface/Electron.h"
 
 namespace PhysicsTools_SelectorUtils {
   struct dictionary {
     pat::strbitset strbitset;
     edm::Wrapper<pat::strbitset> wstrbitset;
     std::vector<pat::strbitset> vstrbitset;
     edm::Wrapper<std::vector<pat::strbitset> > wvstrbitset;
+
+    VersionedSelector<edm::Ptr<reco::Photon> > vPhotonSelector;
+    VersionedSelector<edm::Ptr<pat::Photon> > vPatPhotonSelector;
+    VersionedSelector<edm::Ptr<reco::Muon> > vMuonSelector;
+    VersionedSelector<edm::Ptr<reco::GsfElectron> > vGsfElectronSelector;
+    VersionedSelector<edm::Ptr<pat::Electron> > vPatElectronSelector;
   };
```
This struct is not needed anymore (for a long time).
```suggestion
```

guitargeek replied:
Okay, thanks for the suggestion!

makortel commented on PhysicsTools/SelectorUtils/src/classes_def.xml:
```diff
@@ -9,4 +9,5 @@
   <class name="MuonVPlusJetsIDSelectionFunctor"/>
   <class name="PVSelector"/>
   <class name="RunLumiSelector"/>
+  <class pattern="VersionedSelector<*>"/>
```
The `classes_def.xml` should list all types explicitly (we had not so long ago a campaign to remove all patterns).

makortel commented on RecoEgamma/EgammaTools/interface/AnyMVAEstimatorRun2Base.h:
```diff
@@ -68,8 +74,13 @@ class AnyMVAEstimatorRun2Base {
+  template <class T>
+  class PluginFactory;
+}
 typedef edmplugin::PluginFactory<AnyMVAEstimatorRun2Base*(const edm::ParameterSet&)> AnyMVAEstimatorRun2Factory;
```
To me this sounds more of a reason to move the plugin factory definition out from this file.

guitargeek replied:
Well, I don't see why that's better, but it's also fine for me :) I have now put it in a separate file called `AnyMVAEstimatorRun2Factory.h`. Are you fine with that?

makortel replied:
> Well, I don't see why that's better, but it's also fine for me :)

The users of `AnyMVAEstimatorRun2Base` do not need to know about the factory (only the code creating the object needs the factory). This separation is actually a common pattern in CMSSW (I'd even say vast majority of plugin factories are defined that way, but I didn't do a thorough analysis). The split also avoids the need to explicitly forward-declare types from the framework (code rule 4.7).

> I have now put it in a separate file called `AnyMVAEstimatorRun2Factory.h`. Are you fine with that?

That's good, thanks!

guitargeek replied:
Ok, got it!

makortel commented on RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2.h:
```diff
@@ -47,13 +71,13 @@ class ElectronMVAEstimatorRun2 : public AnyMVAEstimatorRun2Base {
 
   // Data members
-  std::vector<std::unique_ptr<const GBRForest>> gbrForests_;
+  std::vector<const GBRForest*> gbrForests_;
```
Why is the smart pointer removed? (was not clear to me in the commit message)

guitargeek replied:
Because you can't use a `unique_ptr<T>` if T is only forward declared. But that was an over-optimization anyway, I included now `GBRForest.h` again, and generating the dictionaries for the `ElectronMVAEstimatorRun2` is still very fast.

guitargeek commented:
Hi @makortel, thanks for the review! I have addressed your comments, let me know if I should follow up on anything.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/26396

- This PR adds an extra 24KB to repository

cmsbuild commented:
Pull request #35953 was updated. @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @slava77, @jpata can you please check and sign again.

makortel commented on RecoEgamma/EgammaTools/interface/AnyMVAEstimatorRun2Base.h:
```diff
@@ -1,20 +1,26 @@
+#include <vector>
 
-#include "DataFormats/Candidate/interface/Candidate.h"
+namespace edm {
+  class ParameterSet;
+}  // namespace edm
```
Replacing this with
```suggestion
#include "FWCore/Framework/interface/FrameworkfwdMostUsed.h"
```
would be slightly better.

guitargeek replied:
Cool, I didn't know about that!

guitargeek replied:
I guess it's a natural question now why not to use [CandidateFwd.h](https://github.com/cms-sw/cmssw/blob/master/DataFormats/Candidate/interface/CandidateFwd.h) for `reco::Candidate` then, but CandidateFwd.h includes way too much itself, to that's not an option.

perrotta replied:
This comment was not implemented, but it can be deferred to a later clean-up PR

guitargeek replied:
Oh, I thought I addressed it but I think that change got lost in interactive rebasing I guess. I'll make that change another time then!

makortel commented on PhysicsTools/SelectorUtils/python/VIDSelectorBase.py:
```diff
@@ -13,6 +13,34 @@
+    return edm::Ptr<OutPhysObj>(temp);
+  }
+};
+""")
```
Does the dictionary generation get tested in unit tests?

makortel replied:
Unit tests can be added to python files along
```py
if __name__ == "__main__:
    VIDMuonSelector()
```
For a concrete example see e.g. framework python files
https://github.com/cms-sw/cmssw/blob/30bb8dca17f6bc7b7c6b77283f92b9cae3c745b8/FWCore/ParameterSet/python/Config.py#L1762-L1764

In this case I'd add one for each of the `VID(Muon|Electron|Photon)Selector.py` file.

jpata commented:
@cmsbuild please test

(while the other comments are being implemented, we can profit from a test run)

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/20335/summary.html
**COMMIT**: 4e435394e0d8855f5d4241a43d72564fefbfd933
**CMSSW**: CMSSW_12_2_X_2021-11-08-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/20335/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreIntegrationStandalone had ERRORS
---> test unitTestsGroup_1 had ERRORS
---> test unitTestsGroup_4 had ERRORS
---> test testmat had ERRORS
and more ...
</pre>

## RelVals

- [11634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/20335/runTheMatrix-results/11634.911_TTbar_14TeV+2021_DD4hep+TTbar_14TeV_TuneCP5_GenSim+Digi+Reco+HARVEST+ALCA)
```
----- Begin Fatal Exception 08-Nov-2021 19:31:51 CET-----------------------
An exception of category 'PluginNotFound' occurred while
   [0] Constructing the EventProcessor
Exception Message:
Unable to find plugin 'EcalBarrelGeometryEPdd4hep' in category 'CMS EDM Framework ParameterSet Description'. Please check spelling of name.
----- End Fatal Exception -------------------------------------------------
```

jpata commented:
@cmsbuild please test

(to fix .911 which is not relevant)

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/20465/summary.html
**COMMIT**: 4e435394e0d8855f5d4241a43d72564fefbfd933
**CMSSW**: CMSSW_12_2_X_2021-11-10-2300/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/20465/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreIntegrationStandalone had ERRORS
---> test unitTestsGroup_1 had ERRORS
---> test unitTestsGroup_4 had ERRORS
---> test testmat had ERRORS
and more ...
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 2901890
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2901868
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-10-2300+49ad7a/46806/triggerResults): no differences found

jpata commented:
just in case, it looks like the unit test failures are relevant
```
##Failure Location unknown## : Error
Test name: testStandalone::writeAndReadFile
uncaught exception of type std::exception (or derived).
- An exception of category 'FatalRootError' occurred while
   [0] Constructing service of type InitRootHandlers
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::ReadRootmapFile
class  VersionedSelector<edm::Ptr<reco::Muon> > found in libRecoMuonMuonIdentif
ication.so  is already in libPhysicsToolsSelectorUtils.so 
```

jpata commented:
kind ping @guitargeek, just in case

guitargeek commented:
Hi @jpata, sorry for leaving this PR unattended in the last weeks! This week I'm still very busy with other things, but I will come back to this PR next week.

jpata commented:
kind ping on this

jpata commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21128/summary.html
**COMMIT**: 4e435394e0d8855f5d4241a43d72564fefbfd933
**CMSSW**: CMSSW_12_3_X_2021-12-09-1100/slc7_amd64_gcc900
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/21128/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreIntegrationStandalone had ERRORS
---> test unitTestsGroup_1 had ERRORS
---> test unitTestsGroup_4 had ERRORS
---> test testmat had ERRORS
and more ...
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 42
 - DQMHistoTests: Total histograms compared: 3250704
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3250682
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 41 files compared)
 - Checked 177 log files, 37 edm output root files, 42 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2021-12-09-1100+49ad7a/47379/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/27849

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #35953 was updated. @smuzaffar, @Dr15Jones, @makortel, @clacaputo, @cmsbuild, @slava77, @jpata can you please check and sign again.

guitargeek committed:
Add unit test for instantiation of VIDSelectors in Python

guitargeek commented:
Hi @jpata, sorry for following up so late! This is still relevant because https://github.com/cms-sw/cmssw/issues/35754 is still open, right?

As far as I understand there are still two problems here:

1. The one you commented out about the CI failure (https://github.com/cms-sw/cmssw/pull/35953#issuecomment-968982030). I think this is not a problem with the PR, but with the CI incremental build that can't cope with moving dictionaries to another package. This PR completely removes all dictionaries generated in the *Identification packages, so I guess the original file from the master build is not replaces, causing the  `class  VersionedSelector<edm::Ptr<reco::Muon> > found in libRecoMuonMuonIdentification.so  is already in libPhysicsToolsSelectorUtils.so ` error. So it's not a real error and we can ignore it, right? Can you advise, @smuzaffar?
2. Then there is the request for unit tests by @makortel: https://github.com/cms-sw/cmssw/pull/35953#discussion_r742071457. I have added a new commit now where the unit test is implemented.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-35953/27851

- This PR adds an extra 40KB to repository

cmsbuild commented:
Pull request #35953 was updated. @smuzaffar, @Dr15Jones, @makortel, @clacaputo, @cmsbuild, @slava77, @jpata can you please check and sign again.

smuzaffar commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21802/summary.html
**COMMIT**: d4a53bbdc177fd94c3057be16c756c57bde867e3
**CMSSW**: CMSSW_12_3_X_2022-01-18-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/21802/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreIntegrationStandalone had ERRORS
---> test unitTestsGroup_1 had ERRORS
---> test unitTestsGroup_4 had ERRORS
---> test testmat had ERRORS
and more ...
</pre>

## RelVals-INPUT

- [4.76](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21802/runTheMatrixINPUT-results/4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT)```4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT/step2_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT.log```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 43
 - DQMHistoTests: Total histograms compared: 3464860
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3464832
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 42 files compared)
 - Checked 181 log files, 42 edm output root files, 43 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-01-18-2300+49ad7a/47922/triggerResults): no differences found

guitargeek commented:
Since I'm not in CMS anymore I can't log in to check the comparison summary link. Let me know if there is something I should fix!

jpata commented:
Thanks a lot for coming back to help out with this. There are a few unit test errors along the following lines: 

src/FWCore/Integration/test/TestFWCoreIntegrationStandalone
```
===== Test "TestFWCoreIntegrationStandalone" ====
Running .E

##Failure Location unknown## : Error
Test name: testStandalone::writeAndReadFile
uncaught exception of type std::exception (or derived).
- An exception of category 'FatalRootError' occurred while
   [0] Constructing service of type InitRootHandlers
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::ReadRootmapFile
class  VersionedSelector<edm::Ptr<pat::Photon> > found in libRecoEgammaPhotonId
entification.so  is already in libPhysicsToolsSelectorUtils.so 



Failures !!!
Run: 1   Failure total: 1   Failures: 0   Errors: 1

---> test TestFWCoreIntegrationStandalone had ERRORS
TestTime:4
^^^^ End Test TestFWCoreIntegrationStandalone ^^^^
```


src/FWCore/MessageService/test/unitTestsGroup_4
```
===== Test "unitTestsGroup_4" ====
argument 0: unitTestsGroup_4
argument 1: /bin/bash
argument 2: FWCore/MessageService/test
argument 3: u18.sh
argument 4: u19.sh
argument 5: u20.sh
shell is: /bin/bash
Current directory is: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
topdir is: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
testdir is: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/FWCore/MessageService/test
tmpdir is: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/tmp/slc7_amd64_gcc10
testbin is: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/test/slc7_amd64_gcc10
Running script: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/FWCore/MessageService/test/u18.sh
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/tmp/slc7_amd64_gcc10 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Electron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::GsfElectron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Muon> > found in libRecoMuonMuonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
Running script: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/FWCore/MessageService/test/u19.sh
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/tmp/slc7_amd64_gcc10 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Electron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::GsfElectron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Muon> > found in libRecoMuonMuonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
Running script: /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/src/FWCore/MessageService/test/u20.sh
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300/tmp/slc7_amd64_gcc10 /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
0a1,5
> Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
> Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Photon> > found in libRecoEgammaPhotonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
> Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<pat::Electron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
> Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::GsfElectron> > found in libRecoEgammaElectronIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
> Warning in <TInterpreter::ReadRootmapFile>: class  VersionedSelector<edm::Ptr<reco::Muon> > found in libRecoMuonMuonIdentification.so  is already in libPhysicsToolsSelectorUtils.so 
The above discrepancies concern u20_cerr.log
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-18-2300
status = 256

---> test unitTestsGroup_4 had ERRORS
TestTime:15
^^^^ End Test unitTestsGroup_4 ^^^^
```

guitargeek commented:
Ah these errors again, thanks for posting it! As I said already in https://github.com/cms-sw/cmssw/pull/35953#issuecomment-1015992090, I think there are artifacts of the incremental build that we get because the Electron/MuonI/PhotonIdentification package dictionaries are removed in this PR, but in the environment that the CI has they are still there. Probably @smuzaffar can confirm or dismiss this.

smuzaffar commented:
Moving dictionaries should work and CI tests should properly poisoned the deleted dictionaries. Let me check whyc the unit test is failing

smuzaffar commented:
This duplicate dictionary error is coming from `root`. In cmssw full release have [a] and in PR-tests these dicts are now part of RecoEgammaPhotonIdentification_xr.rootmap. As `root` tries to read all `rootmap` files that is why it complains about the duplication. Let me restart the tests with full_cmssw ( hopefully this will avoid the issue) 

[a] RecoEgammaElectronIdentification_xr.rootmap
```
./RecoEgammaPhotonIdentification_xr.rootmap:template <typename T> class VersionedSelector;
./RecoEgammaPhotonIdentification_xr.rootmap:template <typename PhysObj> class MakeVersionedSelector;
./RecoEgammaPhotonIdentification_xr.rootmap:class MakeVersionedSelector<pat::Photon>
./RecoEgammaPhotonIdentification_xr.rootmap:class MakeVersionedSelector<reco::Photon>
./RecoEgammaPhotonIdentification_xr.rootmap:class VersionedSelector<edm::Ptr<pat::Photon> >
./RecoEgammaPhotonIdentification_xr.rootmap:class VersionedSelector<edm::Ptr<reco::Photon> >
```

smuzaffar commented:
test parameters:
- full_cmssw = true

smuzaffar commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21851/summary.html
**COMMIT**: d4a53bbdc177fd94c3057be16c756c57bde867e3
**CMSSW**: CMSSW_12_3_X_2022-01-20-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/21851/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test TestFWCoreIntegrationStandalone had ERRORS
---> test unitTestsGroup_1 had ERRORS
---> test unitTestsGroup_4 had ERRORS
---> test testmat had ERRORS
and more ...
</pre>

## RelVals-INPUT

- [4.76](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21851/runTheMatrixINPUT-results/4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT)```4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT/step2_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT.log```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 43
 - DQMHistoTests: Total histograms compared: 3464860
 - DQMHistoTests: Total failures: 11
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3464827
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 42 files compared)
 - Checked 181 log files, 42 edm output root files, 43 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-01-20-1100+49ad7a/47951/triggerResults): no differences found

smuzaffar commented:
please test with cms-sw/cms-bot#1693

cmsbuild commented:
-1

**Failed Tests**: RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21869/summary.html
**COMMIT**: d4a53bbdc177fd94c3057be16c756c57bde867e3
**CMSSW**: CMSSW_12_3_X_2022-01-20-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/35953/21869/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals-INPUT

- [4.76](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21869/runTheMatrixINPUT-results/4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT)```4.76_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT/step2_ZMuSkim2012D+ZMuSkim2012D+HLTDSKIM2+RECODR1reHLT2+HARVESTDR1reHLT.log```


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 43
 - DQMHistoTests: Total histograms compared: 3464860
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3464833
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 42 files compared)
 - Checked 181 log files, 42 edm output root files, 43 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-01-20-1100+49ad7a/47959/triggerResults): no differences found

jpata commented:
All the unit tests pass, and there are no relevant reco differences. 🚀

makortel commented:
+core

Thanks @guitargeek!

jpata commented:
+reconstruction
- removes unneeded dictionaries related to the electron MVA ID
- no reco changes, no DQM changes
- addresses the dependency issue highlighted in #35754

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1

perrotta commented:
merge

qliphy commented:
@guitargeek  There appear in IB (and also PR tests) some warnings about  duplicated dictionaries or dictionaries defined in the wrong place. Would you please have a look?:

https://cmssdt.cern.ch/SDT/cgi-bin/showDupDict.py/slc7_amd64_gcc10/www/tue/12.3-tue-11/CMSSW_12_3_X_2022-01-25-1100/testLogs/dupDict-lostDefs.log
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-49ad7a/21869/dupDict/lostDefs.txt

Searching for 'classes_def.xml' in '/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_3_X_2022-01-20-1100/src'.

./PhysicsTools/SelectorUtils/src/classes_def.xml
  EgammaCandidates : VersionedSelector< edm::Ptr< reco::Photon > >
  MuonReco : VersionedSelector< edm::Ptr< reco::Muon > >
  EgammaCandidates : VersionedSelector< edm::Ptr< reco::GsfElectron > >

slava77 commented:
> @guitargeek There appear in IB (and also PR tests) some warnings about duplicated dictionaries or dictionaries defined in the wrong place. Would you please have a look?:

CMSSW_12_3_X_2022-01-25-1100 appears to be a patch release. IIUC, we need a full build.

makortel commented:
CMSSW_12_3_X_2022-01-25-1700 is a full build, and the complaint of dictionaries in wrong package is still there
https://cmssdt.cern.ch/SDT/cgi-bin/showDupDict.py/slc7_amd64_gcc10/www/tue/12.3-tue-17/CMSSW_12_3_X_2022-01-25-1700/testLogs/dupDict-lostDefs.log

I think in this case we should update the checker script instead
https://github.com/cms-sw/cmssw/blob/ec200b2e6d0213cdf868308a7a836be8912c4cbc/Utilities/ReleaseScripts/scripts/duplicateReflexLibrarySearch.py#L62
https://github.com/cms-sw/cmssw/blob/ec200b2e6d0213cdf868308a7a836be8912c4cbc/Utilities/ReleaseScripts/scripts/duplicateReflexLibrarySearch.py#L57

In general all dictionaries of DataFormat classes should be declared in the package that defines the class, but in this case `PhysicsTools/SelectorUtils` (that defines `VersionedSelector<T>` already depends on `DataFormats/EgammaCandidates` and `DataFormats/MuonReco`, and therefore it is the correct place to define these three dicrionaries.

@smuzaffar

smuzaffar commented:
@makortel , I will update the dict rule defined here  https://github.com/cms-sw/cmssw/blob/master/Utilities/ReleaseScripts/scripts/duplicateReflexLibrarySearch.py#L32
